File: iot-backend/src/main/java/co/edu/udea/iot/backend/controller/HomeController.java
package co.edu.udea.iot.backend.controller;

import co.edu.udea.iot.backend.model.Home;
import co.edu.udea.iot.backend.service.HomeService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("api/homes")
public class HomeController {

    @Autowired
    private HomeService homeService;

    @Operation(description = "View a list of available homes")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Successfully retrieved list"),
            @ApiResponse(responseCode = "401", description = "You are not authorized to view the resource"),
            @ApiResponse(responseCode = "403", description = "Accessing the resource you were trying to reach is forbidden"),
            @ApiResponse(responseCode = "404", description = "The resource you were trying to reach is not found")
    })
    @GetMapping()
    public List<Home> getAllDevices() {
        return homeService.findAllHomes();
    }

    @Operation(description = "Send a message to a specific device in a specific home")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Successfully retrieved list"),
            @ApiResponse(responseCode = "401", description = "You are not authorized to view the resource"),
            @ApiResponse(responseCode = "403", description = "Accessing the resource you were trying to reach is forbidden"),
            @ApiResponse(responseCode = "404", description = "The resource you were trying to reach is not found")
    })
    @PostMapping("/{homeName}/{deviceName}/{message}")
    public void sendMessage(@PathVariable String homeName, @PathVariable String deviceName, @PathVariable String message) {
        homeService.sendMessage(homeName, deviceName, message);
    }

}

File: backend/smart-home/src/main/kotlin/co/edu/udea/iot/smarthome/controller/HomeController.kt
package co.edu.udea.iot.smarthome.controller

import co.edu.udea.iot.smarthome.dto.HomeDTO
import co.edu.udea.iot.smarthome.facade.HomeFacade
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.media.Content
import io.swagger.v3.oas.annotations.media.Schema
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.responses.ApiResponses
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/homes")
class HomeController(private val homeFacade: HomeFacade) {

    @PostMapping
    @Operation(summary = "Creates a new home record")
    @ApiResponses(value = [
        ApiResponse(responseCode = "200", description = "the home has been successfully created", content = [Content(schema = Schema(implementation = HomeDTO::class))]),
        ApiResponse(responseCode = "400", description = "Invalid request"),
        ApiResponse(responseCode = "500", description = "Internal error")
    ])
    fun createHome(@RequestBody homeDTO: HomeDTO) =
            homeFacade.saveHome(homeDTO)

}
